go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+package vtgate
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+	pbq "github.com/youtube/vitess/go/vt/proto/query"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/tabletserver/tabletconn"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// AddrUpdate updates about an endpoint and whether it is up.
+type AddrUpdate struct {
+	EndPoint *pbt.EndPoint
+	IsUp     bool
+}
+
+// NewHealthCheck creates a new HealthCheck object.
+func NewHealthCheck(endPoints []*pbt.EndPoint, addrUpdate <-chan *AddrUpdate, connTimeout time.Duration, retryDelay time.Duration) *HealthCheck {
+	hc := &HealthCheck{
+		addrToConns: make(map[string]*healthCheckConn),
+		targetToEPs: make(map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint),
+		connTimeout: connTimeout,
+		retryDelay:  retryDelay,
+	}
+	hc.initConns(endPoints)
+	go hc.onAddrUpdate(addrUpdate)
+	return hc
+}
+
+// HealthCheck performs health checking and notifies downstream components about any changes.
+type HealthCheck struct {
+	mu          sync.RWMutex
+	addrToConns map[string]*healthCheckConn                              // addrToConns maps from address to the healthCheckConn object.
+	targetToEPs map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint // targetToEPs maps from keyspace/shard/tablettype to a list of endpoints.
+	connTimeout time.Duration
+	retryDelay  time.Duration
+}
+
+// healthCheckConn contains details about an endpoint.
+type healthCheckConn struct {
+	mu         sync.RWMutex
+	target     *pbq.Target
+	cancelFunc context.CancelFunc
+	conn       tabletconn.TabletConn
+	stats      *pbq.RealtimeStats
+}
+
+// initConns starts health checking for the initial set of endpoints.
+// It waits till all endpoints are tried to connect once.
```

Comment:
I am not sure we need to try to connect to all endpoints once before returning. This may introduce up to connectTimeout delay if one endpoint is down, for no real reason.

I'd rather have a higher level check that makes sure we have enough endpoints before we keep on going...